### PR TITLE
Add type 'ColumnFamilyDescriptor' for 'open' related methods

### PR DESCRIPTION
**Pull request by feb29**
This change breaks API, but makes code more clean and readable too.

### CODE REVIEW

**Comment by siddontang** [comment]
Great work @feb29 

PTAL @BusyJay @huachaohuang  @zhangjinpeng1987

File: src/rocksdb.rs [hunk 1]
```diff
@@ -327,96 +344,89 @@ impl DB {
...
+        let cf_handles: Vec<_> = vec![ptr::null_mut(); cf_names.len()];
+        let cf_options: Vec<_> = options
+            .iter()
+            .map(|x| x.inner as *const crocksdb_ffi::Options)
+            .collect();
+
+        let db = {
+            let db_options = opts.inner;
+            let db_path = cpath.as_ptr();
+            let db_cf_size = cf_names.len() as c_int;
```
Comment (zhangjinpeng1987) [inline src/rocksdb.rs]: s/db_cf_size/db_cfs_count

**Comment by zhangjinpeng1987** [comment]
Nice refactor, rest LGTM

**Comment by feb29** [comment]
rebased to merge https://github.com/pingcap/rust-rocksdb/pull/140

File: src/rocksdb.rs [hunk 2]
```diff
@@ -327,96 +344,89 @@ impl DB {
...
+
+    fn open_cf_internal<'a, T>(opts: DBOptions, path: &str, cfds: Vec<T>) -> Result<DB, String>
+    where
+        T: Into<ColumnFamilyDescriptor<'a>>,
+    {
+        const ERR_CONVERT_PATH: &str = "Failed to convert path to CString when opening rocksdb";
+        const ERR_NULL_DB_ONINIT: &str = "Could not initialize database";
+        const ERR_NULL_CF_HANDLE: &str = "Received null column family handle from DB";
+
+        let cpath = CString::new(path.as_bytes()).map_err(|_| ERR_CONVERT_PATH.to_owned())?;
```
Comment (huachaohuang) [inline src/rocksdb.rs]: @feb29 Hi, we are not using `?` for now, can you use `try!` instead?

File: src/rocksdb.rs [hunk 3]
```diff
@@ -327,96 +344,89 @@ impl DB {
...
-            let cfhandles: Vec<_> = cfs_v.iter().map(|_| ptr::null_mut()).collect();
-
-            let cfopts: Vec<_> = cf_opts_v
-                .iter()
-                .map(|x| x.inner as *const crocksdb_ffi::Options)
-                .collect();
-
-            let db = unsafe {
+            )
+        })?;
```
Comment (huachaohuang) [inline src/rocksdb.rs]: ditto

**Comment by siddontang** [comment]
Thanks @feb29 

LGTM, PTAL @huachaohuang

**Pull request merged**